### PR TITLE
[Pipeline RL] Add support for PipelineRL

### PR DESCRIPTION
**Pull request by jlamypoirier**
This PR provides the initial integration with **PipelineRL** with GRPO loss.

It introduces:

- **Streaming Redis Dataset** — capable of consuming documents such as rollouts from an external Redis stream.
- **Trainer Callback System** — supports callbacks for events like `training_started`, `step_finished`, and `training_finished`.
- **Redis-based Callback Implementation with Weights Broadcast Mechanism** — uses a separate external NCCL rendezvous point to broadcast updated model weights in real time to inference servers, and a Redis stream to broadcast training events.

This enables seamless coordination between Fast-LLM training and PipelineRL-based inference or orchestration components.

### CODE REVIEW

File: fast_llm/data/dataset/streaming.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                # No explicit ACK: messages are processed immediately; on rank failure the job restarts,
+                # so message loss is acceptable and simplifies coordination
+                noack=True,
+            )
+            if messages:
+                for stream_key, msgs in messages:
+                    assert stream_key == REDIS_DATA_STREAM.encode()
+                    for msg_id, msg_data in msgs:
+                        processed += 1
+                        # TODO: or do it after processing all received messaged then count > 1?
```
Comment (bigximik) [inline fast_llm/data/dataset/streaming.py]: We do not need this for now if we stick with consumer groups only. The producer can rely on the group `lag` to control the production rate and on `last-delivered-id` to safely trim the stream.

File: fast_llm/data/dataset/config.py [hunk 2]
```diff
@@ -298,3 +299,56 @@ def build(self, preprocessing: PreprocessingConfig) -> "IndexedDataset[SampleTyp
...
+
+@config_class(dynamic_type={SampledDatasetConfig: "streaming"})
+class StreamingDatasetConfig[SampleType: LanguageModelSample](RedisConfig, SamplableDatasetConfig[SampleType]):
+    """
+    Configuration for a streaming dataset that reads training data from a Redis stream.
+    """
+
+    _abstract = False
+
+    acknowledge_interval: int = Field(
```
Comment (bigximik) [inline fast_llm/data/dataset/config.py]: This is also not needed if we only use consumer gorups, see above comment for implementation.

File: fast_llm/engine/training/streaming.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        if self._do_broadcast:
+            self._client = self._config.get_client()
+            init_method = f"tcp://{config.broadcast.host}:{config.broadcast.port}"
+            logger.info(f"Waiting for weights broadcast rendezvous at {init_method} ...")
+            # TODO: Create a custom process group instead.
+            self._process_group = torch.distributed.init_process_group(
+                backend="nccl",
+                init_method=init_method,
+                world_size=config.broadcast.external_world_size + 1,
+                rank=0,
```
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply fast_llm/engine/training/streaming.py]: We also have control over the external system, we can make it so. It's easier to sync program with hard-coded values than syncing config files.

File: fast_llm/layers/language_model/loss/grpo.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    )
+    per_sample_loss = per_sample_loss * loss_mask
+    loss = per_sample_loss.mean()
+
+    # Sum of per-sequence mean log-probs, matching pipelinerl's new_logprobs metric:
+    #   sum_sum(new_logprobs / num_labels_in_seq, masks_shifted, segments)
+    # Dividing by num_labels_in_seq (span length broadcast per token) and summing over masked
+    # tokens gives mean logprob per sequence; summing those across sequences matches the deepspeed
+    # convention exactly (segments are redundant once num_labels_in_seq is correct).
+    new_logprobs_mean = None if num_labels_in_seq is None else (new_log_probs * loss_mask / num_labels_in_seq).sum()
```
Comment (rafapi) [inline fast_llm/layers/language_model/loss/grpo.py]: here we should do `num_labels_in_seq.clamp(min=1.0)` to avoid division by 0 in cases where complete document is masked

**Review by rafapi — DISMISSED** [review]
